mixer/pkg/cache/lruCache.go

```diff
@@ -0,0 +1,272 @@
+//
+// evictionInterval specifies the frequency at which eviction activities take
+// place. This should be >= to 1 second.
+func NewLRUCache(defaultExpiration time.Duration, evictionInterval time.Duration, maxEntries int) Cache {
+	c := &lruCache{
+		entries:           make([]lruEntry, maxEntries+1),
+		lookup:            make(map[interface{}]int, maxEntries),
+		defaultExpiration: defaultExpiration,
+	}
+
+	// create the linked list of entries
+	for i := 0; i < maxEntries+1; i++ {
+		c.entries[i].next = i + 1
+		c.entries[i].prev = i - 1
+		c.entries[i].expiration = 0x7fffffffffffffff
+	}
+	c.sentinel = &c.entries[0]
+
+	// finish things off, making the list circular
+	c.entries[maxEntries].next = sentinelIndex
+	c.sentinel.prev = maxEntries
+
+	if evictionInterval > 0 {
+		c.baseTime = time.Now().UTC().UnixNano()
+		c.evictionTicker = time.NewTicker(evictionInterval)
+		go c.evicter()
+	}
+
+	return c
+}
+
+func (c *lruCache) Close() {
+	if c.evictionTicker != nil {
+		c.evictionTicker.Stop()
+	}
+}
+
+func (c *lruCache) evicter() {
+	// Wake up once in a while and evict stale items
+	for now := range c.evictionTicker.C {
+		c.evictExpired(now)
+	}
+}
+
+func (c *lruCache) evictExpired(t time.Time) {
+	// We snapshot a base time here such that the time doesn't need to be
+	// sampled in the Set call as calling time.Now() is relatively expensive.
+	// Doing it here provides enough precision for our needs and tends to have
+	// much lower call frequency.
+	n := t.UTC().UnixNano()
+	atomic.StoreInt64(&c.baseTime, n)
+
+	for i := 1; i < len(c.entries); i++ {
+		ent := &c.entries[i]
+
+		if ent.expiration <= n {
+			c.Lock()
+
+			if ent.expiration <= n {
+				c.remove(ent.key)
+			}
+
+			c.Unlock()
+		}
+	}
+}
+
+func (c *lruCache) unlinkEntry(index int) {
+	ent := &c.entries[index]
+
+	c.entries[ent.prev].next = ent.next
+	c.entries[ent.next].prev = ent.prev
+}
+
+func (c *lruCache) linkEntryAtHead(index int) {
+	ent := &c.entries[index]
+
+	ent.next = c.sentinel.next
+	ent.prev = sentinelIndex
+	c.entries[ent.next].prev = index
+	c.sentinel.next = index
+}
+
+func (c *lruCache) linkEntryAtTail(index int) {
+	ent := &c.entries[index]
+
+	ent.next = sentinelIndex
+	ent.prev = c.sentinel.prev
+	c.entries[ent.prev].next = index
+	c.sentinel.prev = index
+}
+
+func (c *lruCache) Set(key interface{}, value interface{}) {
+	c.SetWithExpiration(key, value, c.defaultExpiration)
+}
+
+func (c *lruCache) SetWithExpiration(key interface{}, value interface{}, expiration time.Duration) {
+	exp := atomic.LoadInt64(&c.baseTime) + expiration.Nanoseconds()
+
+	c.Lock()
+
+	index, ok := c.lookup[key]
+	if !ok {
+		index = c.sentinel.prev
+		delete(c.lookup, c.entries[index].key)
+		c.lookup[key] = index
+	}
+
+	c.unlinkEntry(index)
+	c.linkEntryAtHead(index)
+	ent := &c.entries[index]
+	ent.key = key
+	ent.value = value
+	ent.expiration = exp
+
+	c.stats.Writes++
+	c.Unlock()
+}
+
+func (c *lruCache) Get(key interface{}) (interface{}, bool) {
+	c.Lock()
```

Comment:
it'd be cool if we could somehow make `Get` non-mutating, and support a RWMutex, with multiple readers. This might require an alternate mechanism though (like tracking usage counts and maintaining / consulting a separate index on `Set`).

Reply:
The fundamental nature of Get in an LRU cache requires that it does do some mutations to record the usage of the entry.

The MRU idea I mention in the comment at the top of the file is specifically intended to address this problem and increase scalability. By reducing the "mobility" of the most commonly used entries, we eliminate the need for mutation, and thus can avoid serializing on locks and dirtying processor cache lines. You still end up having to do some mutation in order to track usage, but you can do that in a table on the side using atomic ops.

I don't want to invest in this stuff just yet because I'm not quite clear on the actual amount of parallelism this code will be subject to. In the Mixer, we may be juggling a few thousands of requests at a time, not millions. Once these caches are in place, if we find there is in fact contention on the cache locks, we can entertain ideas for more exotic algorithms such as the MRUs.

Reply:
I guess I was wondering if there was a naive-ish approach to `Get` that could be combined with the `go c.evicter()` to do the shuffling in a background routine. Maybe send `index` to `recentlyUsed chan int64` and have a goroutine do the `unlinkEntry` and `linkEntryAtHead` operations. There might be some race condition around completely full caches and infrequently accessed items, but that might be worth it for concurrent reads.

Such an approach _might_ forestall the need more complex MRU operations and indexers, and is fairly straightforward.

Reply:
Oh, that's a very interesting idea. I'll add it to the idea list in the code.